### DISCUSSION TITLE
gtk-fortran on conda-forge

### DISCUSSION THREAD

**Discussion by awvwgk** [vmagnin/gtk-fortran#248] [Ideas]
Have you considered packaging gtk-fortran for conda-forge? GTK3/GTK4 is available on the conda-forge channel, I think there is no plplot at the moment.

**Comment by vmagnin** [comment]
> Have you considered packaging gtk-fortran for conda-forge?

I have never tried to package anything and have no idea of the difficulty of the task. 
What is your experience? Can you post a tutorial that was useful for you so that I can think about it?

**Reply by awvwgk** [reply]
I have packaged a couple of projects for conda-forge (42 and counting). Maybe the one I have staged for plplot would serve as an example: https://github.com/conda-forge/staged-recipes/pull/18106.

**Comment by awvwgk** [comment]
I'll submit a recipe for gtk-fortran in the coming week (maybe already today). Any preferences for packaging? Should I make both GTK3 and GTK4 bindings available (as gtk3-fortran and gtk4-fortran) or just the GTK4 bindings (only gtk-fortran or still gtk4-fortran)?

Another issue might be tests, I show that some are interactive which is bad for testing the package with just ctest. Is there a minimal non-interactive subset for testing?

**Reply by vmagnin** [reply]
@awvwgk your help is very much appreciated!

My opinion is that we should stick to GTK 4 which is now largely available, including conda. And GTK 3 was released in 2011. I am now working only on the gtk4 branch of gtk-fortran (maintaining two branches simultaneously is a pain!).

Concerning tests, gtk-fortran being mainly used for GUI, most of the examples are windows that must be closed with a click. I know that there are frameworks to tests GUI, but it would be probably too much work to use them for gtk-fortran. So I am testing the examples manually for each new GTK version.

But there is also some examples that are not using a GUI and are not interactive:
- build/examples/tests was designed at the beginning of the project to test the ISO_C_BINDING mechanism toward and from GLib/GTK.
- build/examples/regex is calling GLib regular expression functions,
- build/examples/pixbuf_without_gui is calling the GDK Pixbuf library to create a PNG file,
- build/src/gtk-4-fortran is a command to show information about the library and print the GTK version installed on the OS.

If they run correctly, it means GTK is available on the system and Fortran and C bindings are OK. So we can expect that the GUIs will also work.

**Reply by vmagnin** [reply]
The name of the package should be gtk-4-fortran, as defined in the CMakeLists.txt file.

**Reply by awvwgk** [reply]
I'm currently staging a recipe at https://github.com/conda-forge/staged-recipes/pull/18132. If you are interested in becoming a maintainer at conda-forge for gtk-fortran, feel free to comment in this PR and I'll add you.

**Reply by awvwgk** [reply]
That was a difficult one, I got it building on Linux and MacOS eventually, but I gave up on Windows for now.

**Reply by awvwgk** [reply]
@vmagnin Have a look at the invitation at https://github.com/conda-forge/gtk-4-fortran-feedstock/commit/4c62e5dead25446742a0b50aca9b15ce729d61ba#commitcomment-67300083.

**Reply by vmagnin** [reply]
Thanks for your work @awvwgk !
I will have a look at the PR tomorrow (I have already tested it quickly under Ubuntu, and it seemed to be OK).

**Comment by vmagnin** [comment]
I have successfully tested the conda package although I think some problems remain:

```bash
$ conda install gtk-4-fortran
$ whereis gtk-4-fortran
$ export PKG_CONFIG_PATH=/home/osboxes/.conda/envs/fpm/lib/pkgconfig/
$ gfortran julia_pixbuf.f90 $(pkg-config --cflags --libs gtk-4-fortran)
f951: Warning: Nonexistent include directory ‘/home/osboxes/.conda/envs/fpm/include/glib-2.0’ [-Wmissing-include-dirs]
f951: Warning: Nonexistent include directory ‘/home/osboxes/.conda/envs/fpm/lib/glib-2.0/include’ [-Wmissing-include-dirs]
f951: Warning: Nonexistent include directory ‘/home/osboxes/.conda/envs/fpm/include/gio-unix-2.0’ [-Wmissing-include-dirs]
```
Despite the warnings, the examples using the GLib (or gio-unix) functions are running fine, but I guess it is because these libraries were already installed in the system: concerning for example GLib, probably the directory /usr/include/glib-2.0 is used instead of the conda version?

Probably the section "requirements:" in the recipe should be modified? Concerning GLib, should we write `glib` or `libglib`?
https://anaconda.org/conda-forge/libglib

**Reply by awvwgk** [reply]
GLib is splitted in a runtime (`libglib`) and a full version (`glib`). The runtime version is installed by default, if you want to compile against GLib you have to include `glib` in your conda environment as well.

**Reply by vmagnin** [reply]
Does it mean that that line in the requirements section does not install the full version `glib` in the conda environment?
 ```
  host:
    - glib
 ```

**Reply by awvwgk** [reply]
You can find the requirement `glib` imposes when in host in [`run_exports`](https://github.com/conda-forge/glib-feedstock/blob/4972d16722df1d6e58ede39f642ebfc6cab6011d/recipe/meta.yaml#L134-L135). Similar to the `gtk4` package, I'm not splitting `gtk-4-fortran`, meaning there is only the runtime version and you have to explicitly install development dependencies yourself.

We could add `run_exports` for `glib` in `gtk-4-fortran`, however this means dependent packages on conda-forge must explicitly set [`ignore_run_exports`](https://github.com/conda-forge/gtk-4-fortran-feedstock/blob/4c62e5dead25446742a0b50aca9b15ce729d61ba/recipe/meta.yaml#L33-L34) for `glib` if they don't want to pull in the full development dependencies.

**Comment by vmagnin** [comment]
gtk-fortran 4.2.1 is now available:
https://github.com/conda-forge/gtk-4-fortran-feedstock
